> On unix platforms we automatically call __riscv_flush_icache. On non-unix RISC-V dynasm-rs simply doesn't handle the icache invalidation by itself (it is already a no-op, so there shouldn't be a need to stub it out). 

I think this is where we might not be on the same page. While the [full riscv module](https://github.com/CensoredUsername/dynasm-rs/blob/f9a0067cb26bc590905b37b3efc130e56072d588/runtime/src/cache_control.rs#L168-L208) will be compiled in RISC-V, the 2 functions `riscv_flush_icache` are `enforce_ordering_dcache_icache` under different `cfg` flags: `riscv_flush_icache` is only there for `unix` targets, while `enforce_ordering_dcache_icache` always exists.

As a result of this, when we try compiling dynasmrt with a target that is not `unix`, the following error will happen:

```
error[E0425]: cannot find function `riscv_flush_icache` in this scope
   --> runtime/src/cache_control.rs:204:18
    |
204 |             rv = riscv_flush_icache(start, end, flags);
    |                  ^^^^^^^^^^^^^^^^^^ not found in this scope
```

This is the rationale behind this PR. It's really not about a different icache management scheme, but the ability to avoid it when not needed.

Looking at your comments, perhaps it makes more sense to also hide `enforce_ordering_dcache_icache` behind `cfg(unix)` flag?

--- a/runtime/src/cache_control.rs
+++ b/runtime/src/cache_control.rs
@@ -13,7 +13,7 @@ pub fn prepare_for_execution(slice: &[u8]) {
     {
         aarch64::prepare_for_execution()
     }
-    #[cfg(any(target_arch="riscv64", target_arch="riscv32"))]
+    #[cfg(all(unix, any(target_arch="riscv64", target_arch="riscv32")))]
     {
         riscv::enforce_ordering_dcache_icache(slice, true);
     }
@@ -165,7 +165,7 @@ mod aarch64 {
     }
 }
 
-#[cfg(any(target_arch="riscv64", target_arch="riscv32"))]
+#[cfg(all(unix, any(target_arch="riscv64", target_arch="riscv32")))]
 mod riscv {
     // On risc-v, the story about how we synchronize caches is confused.
     // The data sheet states that we ought to do the following.
@@ -185,7 +185,6 @@ mod riscv {
     // we have.
     use std::ffi::{c_void, c_long, c_int};
 
-    #[cfg(unix)]
     extern "C" {
         #[link_name="__riscv_flush_icache"]
         fn riscv_flush_icache(start: *const c_void, end: *const c_void, flags: c_long) -> c_int;

